### DISCUSSION TITLE
Grafana integration with existing prometheus/grafana docker "stack"

### DISCUSSION THREAD

**Discussion by ninjamonkey198206** [amidaware/tacticalrmm#1045] [Q&A] [answered]
The current method looks like it involves a baremetal t-rmm install, using the provided script to install grafana alongside t-rmm.

So far I haven't been able to work through the script to see if I could pull the relevant info to try to create a functional config for t-rmm and grafana running via Docker.

@dinger1986 , it looks like this was your baby. Do you think it's possible to work out a way to do it?

**Comment by ninjamonkey198206** [comment, marked as answer]
I may have found most of what I needed in the script. Will update with progress.

**Reply by ninjamonkey198206** [reply]
Assumes existing and in use Tactical and Prometheus/Grafana stacks. This is to edit both so they can work together without needing to run a separate Grafana instance just for Tactical.

Looks like I've got it mostly working. Used bits from the script, a few searches about attaching to "external" networks, and the walkthrough you referenced: https://community.nethserver.org/t/howto-install-tactical-rmm-including-meshcentral-on-nethserver-with-docker/19112

### Network names need to be edited to suit your install.
### Edit Tactical docker-compose to specify postgres IP address
```text
networks:
  api-db:
    ipv4_address: ${POSTGRES_IP}
```
Either assign the IP manually to what it is currently assigned or use an env variable.

### Edit Grafana docker-compose to add tactical api network
Under network definitions:
```text
tacticalrmm_api-db:
  external: true
```
In grafana service definition
```text
networks:
  monitor-net:
    ipv4_address: ${GRAFANA_IP}
  tacticalrmm_api-db:
```

### Add tacticalrmm data source
Add postgresql data source to Grafana (for now use the tactical user and pass in t-rmm docker-compose config). Edit to suit your configuration, but the Name field MUST be TacticalRMM:

Name: TacticalRMM
Host: postgres-ip:5432
Database: tacticalrmm
User: tactical
Password: postgrespass
Disable TLS/SSL and “Save & test”

![Screenshot 2022-04-03 083232](https://user-images.githubusercontent.com/24654529/161432437-d67bb2a8-1e61-4c89-acd4-5c5f5decc6c1.png)

### Add dashboards to Grafana
Open your Grafana instance in your browser.
Begin importing the new dashboards by copying and pasting the json code for each, or downloading the files and importing the jsons directly.

Only use one set of dashboards, you cannot try both sets at the same time without editing the UIDs for one of the sets.
 
Original Dashboards:

https://github.com/dinger1986/TRMM-Grafana/blob/main/dashboards/mapdash.json
https://github.com/dinger1986/TRMM-Grafana/blob/main/dashboards/sebdash.json
https://github.com/dinger1986/TRMM-Grafana/blob/main/dashboards/tvdash.json

Alternate Dashboards:

https://github.com/dinger1986/TRMM-Grafana/blob/main/dashboards/clientmap.json
https://github.com/dinger1986/TRMM-Grafana/blob/main/dashboards/agentdash.json
https://github.com/dinger1986/TRMM-Grafana/blob/main/dashboards/clientoverview.json

![Screenshot 2022-04-03 102115](https://user-images.githubusercontent.com/24654529/161435055-d0cb80a8-aad9-4baf-9b74-625ab333023e.png)

Leave the UID and names as they are, and import them. Ignore the complaints in my screenshot, I already performed this step.

![Screenshot 2022-04-03 102451](https://user-images.githubusercontent.com/24654529/161435289-9935d9e2-90a1-42fe-b89e-95e91e9dc249.png)

### Enable cpu, disk, and ram checks for the agents you want to monitor.

### Create a URL action for the TacticalRMM Agent dashboard in TacticalRMM Global Settings, only edit the domain to your grafana domain:

Original:
URL Pattern: https://grafana.domain.tld/d/pLkA1-inz/tacticalrmm-agent-dashboard?orgId=1&var-Client={{client.name}}&var-Sites={{site.name}}&var-Agents_HostName={{agent.hostname}}

Alternate:
URL Pattern: https://grafana.domain.tld/d/pLkA1-inz/t-rmm-agent-dashboard?orgId=1&var-Client={{client.name}}&var-Sites={{site.name}}&var-Agents_HostName={{agent.hostname}}

Now you should be able to select a client and run the URL action to open the Grafana T-RMM Agent dashboard and browse to the others via the embedded links.

## Important
When updating T-RMM Docker images, you'll need to stop the Grafana stack first, as it's tied into the T-RMM network, or the T-RMM stack will not properly stop. After updating, bring up T-RMM first, then Grafana.

**Reply by dinger1986** [reply]
Someone did this for neth so instructions should be the same:

https://community.nethserver.org/t/howto-install-tactical-rmm-including-meshcentral-on-nethserver-with-docker/19112

**Reply by ninjamonkey198206** [reply]
It looks like it's similar, just assumes recreating things from scratch.
I'm trying to figure out how to add this into an existing and in use setup with grafana/prometheus as a separate stack being used for other purposes as well. Making decent progress.

**Reply by ninjamonkey198206** [reply]
@dinger1986 Got everything except the map working so far. Not sure if I fully understand the instructions on that part yet.
The reason I wanted to do this is because I've got a separate Prometheus/Grafana stack that gives me info on the docker host, containers, my pfsense, etc, and wanted to integrate this into the existing setup, rather than running two separate instances. Thankfully this needing to be configured as the default datasource doesn't seem to be causing an issue with the other dashboards.

**Reply by dinger1986** [reply]
If you can write up what you have done that would be great.

I havent ever got the map working 100% but I know others have, its just available as examples.

**Reply by ninjamonkey198206** [reply]
@dinger1986 did you want the complete docker-compose files for both stacks, or just the steps I took to make changes? I have those above in the second comment I think. Didn't want to make a huge thread so I just added to it as I went.

**Reply by dinger1986** [reply]
Perfect, shall get a look at it later

**Reply by ninjamonkey198206** [reply]
Works perfectly on my home setup (aside from the map, that's still in progress) but doesn't automatically select the appropriate agent when used from my Google VM. It's likely due to my use of Drop All by default IPTables rules and HAProxy, requiring everything to be bound to localhost rather than publicly exposed.

I'll add an update for that edge case situation once I work it out. It's likely just a tweak to the IPTables rules. Docker networking gets twitchy with that setup, requiring some extra rules to make it play nice.

Edit: Solved. I didn't notice that when I copied, pasted, and edited the URL that it had done something funky and messed it up. Correcting the URL syntax solved the issue.

So, the only quirk left to solve aside from the map is the requirement for it to be the default datasource, as in my case, that gets overridden and the Prometheus datasource takes over every time the stack is cycled. It's a minor annoyance though, as I just have to go into the Tactical datasource and set it back to default.

**Reply by ninjamonkey198206** [reply]
@dinger1986 Adding a datasource variable to the Tactical TRMM dashboard and editing it to select the Tactical postgres datasource seems to eliminate the need for the Tactical datasource to be the default. Unfortunately, I'm having to go through and change each panel to get it to use the variable. I'll fork your repo and upload the modified versions once I'm done.

**Reply by ninjamonkey198206** [reply]
https://github.com/ninjamonkey198206/TRMM-Grafana
The link to the forked repo. Cleaned up some minor typos and changed some of the titles too. Let me know what you think.

**Reply by ninjamonkey198206** [reply]
@dinger1986 I did a thing! Copied your original json files to new files in the repo I forked from yours. Played a bit, changed some titles and descriptions since there are now Linux agents, added the data source variable so the T-RMM data source doesn't need to be the default, moved stuff around so the agent has its own page as does the client/site, and added dashboard links to them all so you can flip back and forth direct from the dashboards!

Woohoo!

I may be a little excited because this is the first time I've edited dashboards, and they actually worked when I was done.

I'm hoping I can add more functions to the agent dashboard, like network traffic, I/O, etc. If you wouldn't mind, I could use some help on the client dashboard, as I duplicated the "client" panel and I'm hoping to have one be for the selected clients, one for sites for the selected clients, and one for agents for the selected clients/sites.

I don't know if it's possible to do securely as it would be cross site, but with the links between the dashboards on the dashboards themselves, maybe the agent panel could be an option to embedded into the bottom section of T-RMM, along with the Summary, Checks, Tasks, and such.

**Reply by dinger1986** [reply]
Sounds really cool, feel free to do a PR, tbh I just facilitated an easy install method for others, Hugo and Sebcashmag on discord are much more knowledgeable about this than I am.
